chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,275 @@
+extern crate lru;
+use lru::LruCache;
+use std::num::NonZeroUsize;
+use std::borrow::Borrow;
+use crate::Signature;
+use crate::hash_to_g2;
+use crate::aggregate_verify as agg_ver;
+use crate::gtelement::GTElement;
+use crate::PublicKey;
+use std::collections::HashMap;
+use sha2::{Digest, Sha256};
+
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyList, PyBool};
+
+pub type Bytes32 = [u8; 32];
+pub type Bytes48 = [u8; 48];
+
+#[cfg_attr(
+    feature = "py-bindings",
+    pyclass(name = "BLSCache"),
+)]
+pub struct BLSCache {
+    cache: LruCache<Bytes32, GTElement>,
+}
+
+impl BLSCache {
+    
+    pub fn generator(cache_size: Option<usize>) -> Self {
+        let cache: LruCache<Bytes32, GTElement> = LruCache::new(NonZeroUsize::new(cache_size.unwrap_or(50000)).unwrap());
+        Self{cache}
+    }
+    
+    pub fn get_pairings<P: Borrow<[Bytes48]>, M: Borrow<[Vec<u8>]>> (
+        &mut self,
+        pks: &P,
+        msgs: &M,
+        force_cache: bool,
+    ) -> Vec<GTElement> {
+        let mut pairings: Vec<Option<GTElement>> = vec![];
+        let mut missing_count: usize = 0;
+        
+        for (pk, msg) in pks.borrow().iter().zip(msgs.borrow().iter()) {
+            let mut aug_msg = pk.to_vec();
+            aug_msg.extend_from_slice(msg.borrow()); // pk + msg
+            let mut hasher = Sha256::new();
+            hasher.update(aug_msg);
+            let h: Bytes32 = hasher.finalize().into();
+            let pairing: Option<&GTElement> = self.cache.get(&h);
+            match pairing {
+                Some(pairing) => {
+                    if !force_cache {
+                        // Heuristic to avoid more expensive sig validation with pairing
+                        // cache when it's empty and cached pairings won't be useful later
+                        // (e.g. while syncing)
+                        missing_count += 1;
+                        if missing_count > pks.borrow().len() / 2 {
+                            return vec![];
+                        }
+                    }
+                    pairings.push(Some(pairing.clone()));
+                },
+                _ => {
+                    pairings.push(None);
+                },
+            }
+            
+        }
+
+        // G1Element.from_bytes can be expensive due to subgroup check, so we avoid recomputing it with this cache
```

Comment:
Not sure why there are two loops instead of one. Seems like a natural way to do this would be do dig for the answer in the cache, and if it ain't there, calculate it (and possible cache the result). Maybe it has something to do with the heuristic above. I realize this is just a port so you may not know either.